src/openms/source/ANALYSIS/OPENSWATH/OpenSwathHelper.cpp

```diff
@@ -162,6 +165,120 @@ namespace OpenMS
     return std::make_pair(min,max);
   }
 
+  OpenSwath::LightTargetedExperiment
+  OpenSwathHelper::sampleExperiment(
+    const OpenSwath::LightTargetedExperiment & exp,
+    Size bins,
+    Size peptides_per_bin,
+    unsigned int seed,
+    bool filter_decoys,
+    bool sort_by_intensity,
+    double top_fraction)
+  {
+    // 0) initial candidate selection: exclude decoys if requested
+    std::vector<OpenSwath::LightCompound> candidates;
+    if (filter_decoys)
+    {
+      std::unordered_set<String> good_ids;
+      for (auto & tr : exp.getTransitions())
+      {
+        if (!tr.decoy)
+          good_ids.insert(tr.getPeptideRef());
+      }
+      for (auto & cmp : exp.getCompounds())
+      {
+        if (good_ids.count(cmp.id))
+          candidates.push_back(cmp);
+      }
+    }
+    else
+    {
+      candidates = exp.getCompounds();
+    }
+
+    // 1) optionally sort by library intensities and trim to top fraction
+    if (sort_by_intensity && top_fraction > 0.0 && top_fraction <= 1.0)
+    {
+      // sum intensities per peptide across all transitions
+      std::unordered_map<String, double> intensity_sum;
+      for (auto & tr : exp.getTransitions())
+      {
+        if (!filter_decoys || !tr.decoy)
+        {
+          intensity_sum[tr.getPeptideRef()] += tr.library_intensity;
+        }
+      }
+
+      // sort candidates by descending sum
+      std::sort(candidates.begin(), candidates.end(), [&](auto & a, auto & b) {
+        return intensity_sum[a.id] > intensity_sum[b.id];
+      });
+
+      // trim to top N%
+      Size max_keep = std::max<Size>(1, static_cast<Size>(candidates.size() * top_fraction));
+      candidates.resize(std::min(max_keep, candidates.size()));
+    }
+
+    // 2) estimate RT range using filtered candidates
+    double rt_min = std::numeric_limits<double>::max();
+    double rt_max = std::numeric_limits<double>::lowest();
+    for (auto & cmp : candidates)
+    {
+      rt_min = std::min(rt_min, cmp.rt);
+      rt_max = std::max(rt_max, cmp.rt);
+    }
+    double bin_width = (rt_max - rt_min) / static_cast<double>(bins);
+
+    // 3) sample uniformly across RT bins
+    std::vector<OpenSwath::LightCompound> picked;
+    std::mt19937 rng;
+    rng.seed(seed == 0 ? std::random_device{}() : seed);
```

Comment:
@timosachsenberg is this the best way to do this? Does OpenMS have an internal random number generator that is preferred?

Reply:
IIRC we have used boost for rngs in the past to make sure that the implementation is stable across different OS/std libs.

Reply:
yeah, unfortunately the STL ones are not portable (different sequences on different platforms) 
We use boost https://github.com/OpenMS/OpenMS/blob/develop/src/openms/include/OpenMS/MATH/MathFunctions.h#L470